src/stream/src/executor/top_n_new.rs

```diff
@@ -205,171 +205,47 @@ impl<S: StateStore> TopNExecutorBase for InnerTopNExecutorNew<S> {
-                        if ordered_pk_row < *start_key {
-                            // deleted row in [0, offset) the topn range will move forward
-
-                            // update both ends
-                            if start_row.ordered_key != new_start_row.ordered_key {
-                                new_ops.push(Op::Delete);
-                                new_rows.push(start_row.row);
-                            }
-
-                            if new_end_row.is_valid()
-                                && end_row.ordered_key != new_end_row.ordered_key
-                            {
-                                new_ops.push(Op::Insert);
-                                new_rows.push(new_end_row.row);
-
-                                if new_end_row.ordered_key < end_row.ordered_key {
-                                    new_ops.push(Op::Delete);
-                                    new_rows.push(end_row.row);
-                                }
-                            }
-                        } else {
-                            let mut is_contained = end_row.ordered_key.is_none();
-                            if let Some(end_key) = end_row.ordered_key.as_ref() {
-                                is_contained = ordered_pk_row <= *end_key;
-                            }
-                            // deleted row in (offset+limit, Inf)
-                            if !is_contained {
-                                continue;
-                            }
-
-                            // deleted row in [offset, offset+limit], emit change of result set
-                            new_ops.push(Op::Delete);
-                            new_rows.push(row);
-
-                            // update high end if nedded
-                            if new_end_row.is_valid()
-                                && new_end_row.ordered_key != end_row.ordered_key
-                            {
-                                new_ops.push(Op::Insert);
-                                new_rows.push(new_end_row.row);
-                                if end_row.is_valid()
-                                    && new_end_row.ordered_key < end_row.ordered_key
-                                {
-                                    new_ops.push(Op::Delete);
-                                    new_rows.push(end_row.row);
-                                    assert!(end_row.ordered_key.is_some());
-                                }
-                            }
-                        }
-                    }
+                        .delete(&ordered_pk_row, row.clone(), epoch)?;
                 }
             }
         }
-        generate_output(new_rows, new_ops, &self.schema)
+
+        // find the top-n range after modification
+        let new_rows = self
+            .managed_state
+            .find_range(self.offset, num_limit, epoch)
+            .await?;
+
+        // delete previous result set
```

Comment:
May add a TODO here.

TODO: only emit the difference instead of emitting the full contents.

Reply:
+1.